tgui/packages/tgui/interfaces/space_heater.js

```diff
@@ -0,0 +1,275 @@
+          verticalAlign={"middle"}>
+          <ProgressBar
+            grow
+            color={Set_Color(set_temperature, "green", on)}
+            ranges={{
+              "green": [0.5, Infinity],
+              "yellow": [0.1, 0.5],
+              "red": [-Infinity, 0.1],
+            }}
+            value={Math.max(0, [cell !== null ? cell_charge/100 : 0])}
+          />
+        </LabeledList.Item>
+      </LabeledList>
+    </Section>
+  );
+};
+
+const TemperatureRegulator = (props, context) => {
+  const { data, act } = useBackend(context);
+  const {
+    emagged,
+    on,
+    min,
+    max,
+    cell,
+    cell_charge,
+    set_temperature,
+  } = data;
+  return (
+    <Section title={Glitch_Text(emagged, "Temperature regulator", 3)}>
+      {!!on && !!emagged && !HVAC_Death(emagged, cell, cell_charge) && (
+        <Modal
+          backgroundColor={Set_Color(set_temperature, "#001414")}
+          align="center"
+          height={5.8}
+          width={28.5} // This sucks
+          mr={1.4}
+        >
+          {Generate_Emag_Text(Find_Theme(emagged, set_temperature, on), 5)}
+        </Modal>
+      )}
+      <Stack justify="center" align="center">
+        <Stack.Item>
+          <Button
+            icon={"fast-backward"}
+            tooltip={"Set minimum"}
+            disabled={set_temperature === min}
+            onClick={() => act("set_temp", { inputted_temperature: min })} />
+        </Stack.Item>
+        <Stack.Item>
+          <Button
+            icon={"backward"}
+            tooltip={emagged ? "Decrease by 50": "Decrease by 5"}
+            disabled={set_temperature === min}
+            onClick={() => act("set_temp", { temperature_adjust: emagged ? -50: -5 })} />
+        </Stack.Item>
+        <Stack.Item>
+          <Button
+            icon={"equals"}
+            tooltip={"Room temperature"}
+            onClick={() => act("set_temp", { inputted_temperature: 293 })} />
+        </Stack.Item>
+        <Stack.Item>
+          <Button
+            icon={"forward"}
+            tooltip={emagged ? "Increase by 50": "Increase by 5"}
+            disabled={set_temperature === max}
+            onClick={() => act("set_temp", { temperature_adjust: emagged ? 50: 5 })} />
+        </Stack.Item>
+        <Stack.Item>
+          <Button
+            icon={"fast-forward"}
+            tooltip={"Set maximum"}
+            disabled={set_temperature === max}
+            onClick={() => act("set_temp", { inputted_temperature: max })} />
+        </Stack.Item>
+      </Stack>
+      <Divider />
+      <Slider
+        value={set_temperature}
+        format={value => value+" K"}
+        bold
+        textColor={set_temperature >= 293 ? "red": "blue"}
+        minValue={min}
+        maxValue={max}
+        step={emagged ? 5: 1}
+        stepPixelSize={emagged ? 1.5: 4.5}
+        ranges={{
+          "blue": [-Infinity, 292],
+          "red": [293, Infinity],
+        }}
```

Comment:
I had looked into temperatureUtils.js before and tried to use getTemperatureColor, but I had trouble using it successfully.  If I remember correctly it was about the minimum temperature in its non-emagged state, will poke around with it tommorow since its most definitely much more preferable over the sudden change in color with the slider.

The constants would also definitely be more preferable over the magic numbers, will get on that tommorow.

Reply:
I have added NeutralTemperature and replaced 293 with it, as well in the spaceheater.dm replaced 293 with T20C and added it to max,min definitions. The magic numbers of 400 and 100 used to figure out when the special emag variations start have been left alone. I would use T100C (boiling temperature of water) in place of 400, but I dont have any idea what would make sense in-place of 100, so I prefer to keep them both arbitrarily decided rather than give one meaning while the other stays without.

With the getTemperatureColor I dont think I can add it, because of the inconsistencies that would arise from non-emagged and emagged. Emagged would work fine to add it, but then I am also required to add it to non-emagged which is where a problem comes. I cant set the minimum temperature for the function, which means 258 the lowest possible temp when non-emagged, is considered a warm color when it should be considered the coldest. Could be fixed by updating the function to allow to set the minimum, altough I really dont understand the math thats going on with it and was unable to figure it out.